File: swift/Sources/ImogenSDKConformance/Stub.swift
import Foundation

#if canImport(FoundationNetworking)
    import FoundationNetworking
#endif

/// A stub imogen, as a `URLProtocol`.
///
/// The conformance suite needs a server that records what it was asked and answers
/// predictably. Intercepting `URLSession` gets that without opening a socket, which keeps
/// the tests free of anything the library itself does not already depend on.

struct Recorded: Sendable {
    let method: String
    let path: String
    let query: String
    let headers: [String: String]
    let body: Data
}

struct Reply: Sendable {
    var status: Int = 200
    var body: String = "{}"

    static func json(_ body: String) -> Reply { Reply(status: 200, body: body) }
    static func status(_ status: Int, _ body: String) -> Reply { Reply(status: status, body: body) }
}

/// Shared because `URLProtocol` instances are made by the loading system, not by us.
final class StubState: @unchecked Sendable {
    private let lock = NSLock()
    private var recorded: [Recorded] = []
    private var responder: (@Sendable (Recorded, Int) -> Reply)?

    static let shared = StubState()

    func install(_ responder: @escaping @Sendable (Recorded, Int) -> Reply) {
        lock.lock()
        defer { lock.unlock() }
        self.responder = responder
        self.recorded = []
    }

    func answer(_ request: Recorded) -> Reply {
        lock.lock()
        let index = recorded.count
        recorded.append(request)
        let responder = self.responder
        lock.unlock()
        return responder?(request, index) ?? Reply()
    }

    var calls: [Recorded] {
        lock.lock()
        defer { lock.unlock() }
        return recorded
    }

    var callCount: Int { calls.count }
}

final class StubProtocol: URLProtocol {
    override class func canInit(with request: URLRequest) -> Bool { true }

    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }

    override func startLoading() {
        let url = request.url!
        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)

        // `httpBody` is nil when URLSession turned the body into a stream, which it does
        // for anything of any size, so read it back rather than trusting the property.
        var body = request.httpBody ?? Data()
        if body.isEmpty, let stream = request.httpBodyStream {
            stream.open()
            defer { stream.close() }
            var buffer = [UInt8](repeating: 0, count: 64 * 1024)
            while stream.hasBytesAvailable {
                let read = stream.read(&buffer, maxLength: buffer.count)
                if read <= 0 { break }
                body.append(contentsOf: buffer[0..<read])
            }
        }

        let recorded = Recorded(
            method: request.httpMethod ?? "GET",
            path: components?.path ?? url.path,
            query: components?.query ?? "",
            headers: Dictionary(
                (request.allHTTPHeaderFields ?? [:]).map { ($0.key.lowercased(), $0.value) },
                uniquingKeysWith: { first, _ in first }
            ),
            body: body
        )

        let reply = StubState.shared.answer(recorded)
        let response = HTTPURLResponse(
            url: url,
            statusCode: reply.status,
            httpVersion: "HTTP/1.1",
            headerFields: ["Content-Type": "application/json"]
        )!

        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
        client?.urlProtocol(self, didLoad: Data(reply.body.utf8))
        client?.urlProtocolDidFinishLoading(self)
    }

    override func stopLoading() {}
}

/// A session wired to the stub, plus a fresh recording.
func stubbedSession(_ responder: @escaping @Sendable (Recorded, Int) -> Reply) -> URLSession {
    StubState.shared.install(responder)
    let configuration = URLSessionConfiguration.ephemeral
    configuration.protocolClasses = [StubProtocol.self]
    return URLSession(configuration: configuration)
}

File: kotlin/src/test/kotlin/com/imogen/sdk/Stub.kt
package com.imogen.sdk

import io.ktor.client.HttpClient as KtorClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.client.request.HttpRequestData
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.utils.io.ByteChannel
import io.ktor.utils.io.readRemaining
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.io.readByteArray

/**
 * A stub imogen, built on ktor's MockEngine.
 *
 * The conformance suite needs a server that records what it was asked and answers
 * predictably. Intercepting the client's engine gets that without opening a socket, which
 * keeps the tests free of anything the library itself does not already depend on.
 */

data class Recorded(
    val method: String,
    val path: String,
    val query: String,
    val headers: Map<String, String>,
    val body: ByteArray,
) {
    // Data classes with an array member need these written out to compare sensibly.
    override fun equals(other: Any?): Boolean = this === other

    override fun hashCode(): Int = System.identityHashCode(this)
}

data class Reply(val status: Int = 200, val body: String = "{}") {
    companion object {
        fun json(body: String) = Reply(200, body)
        fun status(status: Int, body: String) = Reply(status, body)
    }
}

class Stub(private val responder: (Recorded, Int) -> Reply) {
    val calls: MutableList<Recorded> = mutableListOf()

    val callCount: Int get() = calls.size

    val engine: KtorClient = KtorClient(
        MockEngine { request ->
            val recorded = record(request)
            val index = synchronized(calls) {
                calls.add(recorded)
                calls.size - 1
            }
            val reply = responder(recorded, index)

            respond(
                content = reply.body,
                status = HttpStatusCode.fromValue(reply.status),
                headers = headersOf(HttpHeaders.ContentType, ContentType.Application.Json.toString()),
            )
        }
    )

    private suspend fun record(request: HttpRequestData): Recorded {
        val body = runCatching {
            request.body.toByteArrayOrEmpty()
        }.getOrDefault(ByteArray(0))

        return Recorded(
            method = request.method.value,
            path = request.url.encodedPath,
            query = request.url.encodedQuery,
            headers = request.headers.entries().associate { it.key.lowercase() to it.value.first() },
            body = body,
        )
    }
}

private suspend fun io.ktor.http.content.OutgoingContent.toByteArrayOrEmpty(): ByteArray =
    when (this) {
        is io.ktor.http.content.OutgoingContent.ByteArrayContent -> bytes()
        is io.ktor.http.content.OutgoingContent.ReadChannelContent ->
            readFrom().readRemaining().readByteArray()
        // A multipart form is written rather than read, so it needs somewhere to be
        // written to before there is anything to record. Concurrently, because a body
        // larger than the channel's buffer would otherwise deadlock against itself.
        is io.ktor.http.content.OutgoingContent.WriteChannelContent -> coroutineScope {
            val channel = ByteChannel()
            launch {
                writeTo(channel)
                channel.flushAndClose()
            }
            channel.readRemaining().readByteArray()
        }
        else -> ByteArray(0)
    }
